extensions/replication/queueProcessor/QueueProcessor.js

```diff
@@ -221,48 +223,66 @@ class QueueProcessor {
 
     _getData(entry, cb) {
         this.log.debug('getting data', { entry: entry.getLogInfo() });
-        const req = this.S3source.getObject({
-            Bucket: entry.getBucket(),
-            Key: entry.getObjectKey(),
-            VersionId: entry.getEncodedVersionId() });
-        const incomingMsg = req.createReadStream();
-        req.on('error', err => {
-            this.log.error('error response getting data from S3',
-                           { entry: entry.getLogInfo(),
-                             origin: this.sourceConfig.s3,
-                             error: err.message,
-                             httpStatus: err.statusCode });
-            incomingMsg.emit('error', err);
+        const locations = entry.getLocation();
+        return async.timesLimit(locations.length, LIMIT, (i, next) => {
```

Comment:
You are right, that would be a nice optimization. I think `async.queue` is a good fit here. You can keep adding streams to the queue, set the concurrency of how many you want to process in parallel and the "processor" is invoked as soon as something is added to the queue.